scripts/zkai_native_attention_mlp_adapter_opening_geometry_budget_gate.py

```
@@ -0,0 +1,823 @@
+
+def ratio(numerator: int, denominator: int) -> float | None:
+    if denominator <= 0:
+        return None
+    return round(numerator / denominator, 6)
+
+
+def read_json_and_raw(path: pathlib.Path) -> tuple[Any, bytes]:
+    fd: int | None = None
+    try:
+        if path.is_symlink():
+            raise AdapterOpeningGeometryBudgetError(f"refusing to read symlink: {path}")
```

Ruff (0.15.12)

[warning] Avoid specifying long messages outside the exception class

(TRY003)